src/main/java/org/myrobotlab/service/Ros.java

```diff
@@ -1,38 +1,316 @@
       Ros ros = (Ros) Runtime.start("ros", "Ros");
-      ros.startService();
-      /*
-       * SwingGui gui = new SwingGui("gui"); gui.startService();
-       */
+
+      WebGui webgui = (WebGui) Runtime.create("webgui", "WebGui");
+      // webgui.setSsl(true);
+      webgui.autoStartBrowser(false);
+      webgui.setPort(8888);
+      // webgui.setSsl(true);
+      webgui.startService();
 
     } catch (Exception e) {
-      Logging.logError(e);
+      log.error("main threw", e);
     }
   }
 
+  Map<String, RosServiceCallback> callbacks = new HashMap<>();
+
+  @SuppressWarnings("rawtypes")
+  transient private WsClient client = null;
+
+  protected boolean connected = false;
+
   public Ros(String n, String id) {
     super(n, id);
   }
 
-  static public String[] getCategories() {
-    return new String[] { "bridge" };
+  @Override
+  public ServiceConfig apply(ServiceConfig c) {
+    RosConfig config = (RosConfig) super.apply(c);
+    if (config.connect) {
+      connect(config.bridgeUrl);
+      if (config.subscriptions != null) {
+        for (String topic : config.subscriptions) {
+          rosSubscribe(topic);
+        }
+      }
+    }
+    return c;
+  }
+
+  // FIXME - TODO reconnect
+  public void connect(String url) {
+
+    if (connected) {
+      info("already connected");
+      return;
+    }
+
+    client = new WsClient();
+    client.connect(this, url);
+  }
+
+  // TODO - getTopics
+  // getSubscriptions
+  // setNodeName ?
+  //
+
+  public void disconnect() {
+    if (client != null) {
+      client.close();
+    }
+  }
+
+  @Override
+  public ServiceConfig getConfig() {
+    RosConfig config = (RosConfig) super.getConfig();
+    config.connect = connected;
+    return config;
+  }
+
+  public List<String> getSubscriptions() {
+    return ((RosConfig) config).subscriptions;
+  }
+
+  public List<String> getTopics() throws InterruptedException {
+
+    RosServiceCallback ret = rosCallService("/rosapi/topics");
+    if (ret == null) {
+      return null;
+    }
+    RosMsg msg = (RosMsg) ret.msg;
+    if (msg != null) {
+      List topics = (List) ((Map) msg.values).get("topics");
+      log.info(ret.toString());
+      return topics;
+    } else {
+      return new ArrayList<>();
+    }
+  }
+
+  public RosMsg publishRosMsg(RosMsg msg) {
+    return msg;
+  }
+
+  // FIXME - its a blocking call
+  public RosServiceCallback rosCallService(String service) {
+    try {
+      RosConfig c = (RosConfig) config;
+      RosServiceCallback callback = new RosServiceCallback();
+      String id = UUID.randomUUID().toString();
+      callback.id = id;
+      callbacks.put(id, callback);
+
+      RosMsg msg = new RosMsg();
+      msg.id = id;
+      msg.op = "call_service";
+      msg.service = service;
+
+      sendJson(CodecUtils.toJson(msg));
+      synchronized (callback) {
+        callback.wait(c.serviceCallTimeoutMs);
+      }
+      return callback; // callbacks.get(id);
+
+    } catch (TimeExceededException ex) {

```

Comment:
This `TimeExceeededException` is from Apache Lucene, I don't think that's what you meant to put here. As far as I can tell the `wait()` call doesn't throw an exception at all if it times out, instead you need to check the time yourself. From the official API docs, this is the recommended pattern:

```java
synchronized (obj) {
    while (<condition does not hold> && <timeout not exceeded>) {
          long timeoutMillis = ... ; // recompute timeout values
          int nanos = ... ;
          obj.wait(timeoutMillis, nanos);
      }      ... // Perform action appropriate to condition or timeout  
```

Reply:
Heh, thanks I most certainly did not

Reply:
in your example though .. there is already a timeout on obj.wait ... my intention is only to use that

Reply:
Unfortunately, the timeout does not guarantee that the thread will not be woken up by the operating system for a different reason, all that timeout does is tell the scheduler to wake up the thread if the timeout has been exceeded and the thread has not been woken up by any other cause. The while loop ensures that we only exit the block once we are sure we have a valid message or are sure the timeout has passed

Reply:
:thinking: .. been a while since i've look at this .. but 
"the timeout does not guarantee that the thread will not be woken up by the operating system for a different reason" <- for what? ...  I looked at the callback processing its doing a notifyAll() but doing it on the synchronized "specific" callback.
There should be no other notifies for this object, if a timeout is reached it should return with a null msg

Reply:
This problem is lower than Java `notify()` calls. The wait() is implemented using an operating system thread conditional, which can be prematurely awoken on a couple of different conditions. An example of one such condition on Linux is described [here](https://stackoverflow.com/a/1051816)

Reply:
NOOOOOOO ! :facepalm: 
Thanks AP ... I'll put in the #*X#&@ loop

Reply:
fixed .. also found out a change between melodic and noetic service calls - noetic the compression field is required